src/sentry/snuba/metrics/datasource.py

```diff
@@ -139,55 +146,46 @@ def get_available_derived_metrics(
     return found_derived_metrics.intersection(public_derived_metrics)
 
 
-def get_metrics(projects: Sequence[Project], use_case_id: UseCaseID) -> Sequence[MetricMeta]:
-    ENTITY_TO_DATASET = {
-        "sessions": {
-            "c": "metrics_counters",
-            "s": "metrics_sets",
-            "d": "metrics_distributions",
-            "g": "metrics_gauges",
-        },
-        "transactions": {
-            "c": "generic_metrics_counters",
-            "s": "generic_metrics_sets",
-            "d": "generic_metrics_distributions",
-            "g": "generic_metrics_gauges",
-        },
-        "spans": {
-            "c": "generic_metrics_counters",
-            "s": "generic_metrics_sets",
-            "d": "generic_metrics_distributions",
-            "g": "generic_metrics_gauges",
-        },
-        "custom": {
-            "c": "generic_metrics_counters",
-            "s": "generic_metrics_sets",
-            "d": "generic_metrics_distributions",
-            "g": "generic_metrics_gauges",
-        },
-    }
+def get_metrics_meta(projects: Sequence[Project], use_case_id: UseCaseID) -> Sequence[MetricMeta]:
+    metas = []
+    unique_mris = set(get_known_mris(use_case_id) + get_stored_mris(projects, use_case_id))
 
-    result = []
-    for mri in get_all_mris():
-        parsed = parse_mri(mri)
+    for mri in sorted(unique_mris):
+        parsed_mri = parse_mri(mri)
 
-        if parsed.entity not in ENTITY_TO_DATASET["sessions"].keys():
-            ops = []
-        elif parsed.namespace == "sessions":
-            ops = AVAILABLE_OPERATIONS[ENTITY_TO_DATASET[parsed.namespace][parsed.entity]]
-        else:
-            ops = AVAILABLE_GENERIC_OPERATIONS[ENTITY_TO_DATASET[parsed.namespace][parsed.entity]]
+        ops = get_available_operations(parsed_mri)
 
-        result.append(
+        metas.append(
             MetricMeta(
                 mri=mri,
-                unit=parsed.unit,
-                name=parsed.name,
+                name=parsed_mri.name,
                 operations=ops,
+                # TODO: check unit casting
+                unit=parsed_mri.unit,
+                # TODO: add entity letter to entity key mapping i.e. c -> counter
+                type=parsed_mri.entity,
             )
         )
 
-    return result
+    return metas
+
+
+def get_stored_mris(projects: Sequence[Project], use_case_id: UseCaseID) -> Sequence[str]:
```

Comment:
I would suggest having the function's return value be of type `List[str]` since the `get_known_mris` function also uses that. In both, you return the concrete `list` type.